File: SquidstatLibrary/examples/Python/dataOutput.py
"""! @example dataOutput.py """
import sys
from PySide6.QtWidgets import QApplication
from PySide6.QtCore import  QTextStream, QFile, QStandardPaths, QIODevice
from SquidstatPyLibrary import AisDeviceTracker, AisErrorCode
from SquidstatPyLibrary import AisExperiment
from SquidstatPyLibrary import AisInstrumentHandler
from SquidstatPyLibrary import AisConstantPotElement
from SquidstatPyLibrary import AisConstantCurrentElement

# Define relavant device information, for easy access
COMPORT = "COM1"
CHANNEL = 0

app = QApplication()

tracker = AisDeviceTracker.Instance()

cvElement = AisConstantPotElement(1, 1, 30)
ccElement = AisConstantCurrentElement(0.001, 1, 60)

experiment = AisExperiment()

success = True

success &= experiment.appendElement(cvElement, 1)
success &= experiment.appendElement(ccElement, 1)

if not success:
    print("Error building experiment")
    sys.exit()

filePath = ""
fileNum = 1

def onNewElementStarting(channel, info):
    name = f"/{fileNum}_{info.stepNumber}_{info.stepName}.csv"
    filePath = QStandardPaths.writableLocation(QStandardPaths.DesktopLocation) + name

    file = QFile(filePath)
    if not file.open(QIODevice.WriteOnly | QIODevice.Text):
        return

    # Writing headers to file
    out = QTextStream(file)
    out << "Time Stamp, Counter Electrode Voltage, Working Electrode Voltage, Current \n"
    file.close()

    print(f"New element beginning: {info.stepName} step: {info.stepNumber}")

def onActiveDCDataReady(channel, data):
    print(f"current: {data.current}  voltage: {data.workingElectrodeVoltage}  counter electrode: {data.counterElectrodeVoltage}  timestamp: {data.timestamp}")

    # Save the DC data to the file created at element beginning
    file = QFile(filePath)
    if not file.open(QIODevice.Append | QIODevice.WriteOnly | QIODevice.Text):
        return
    
    out = QTextStream(file)
    out << data.timestamp << ","<< data.counterElectrodeVoltage << "," << data.workingElectrodeVoltage << "," << data.current << "\n"
    file.close()

def connectSignals(handler):  
    handler.activeDCDataReady.connect(onActiveDCDataReady)
    handler.activeACDataReady.connect(lambda channel, data: print(f"Timestamp: {data.timestamp} Frequency: {data.frequency} Absolute Impedance: {data.absoluteImpedance}"))
    handler.experimentNewElementStarting.connect(onNewElementStarting)
    handler.experimentStopped.connect(lambda channel, reason: (print(f"Experiment Stopped Signal {channel}, {reason}"), app.quit()))
    handler.deviceError.connect(lambda channel, error: print(f"Device Error: {error}"))

def runExperiment(deviceName):
    handler = tracker.getInstrumentHandler(deviceName)
    
    connectSignals(handler)

    error = handler.uploadExperimentToChannel(CHANNEL, experiment)
    # Exit the application if there is any error uploading experiment
    if error.value() != AisErrorCode.Success:
        print(error.message())
        app.quit()

    error = handler.startUploadedExperiment(CHANNEL)
    # Exit the application if there is any error starting experiment
    if error.value() != AisErrorCode.Success:
        print(error.message())
        app.quit()

tracker.newDeviceConnected.connect(runExperiment)
tracker.deviceDisconnected.connect(lambda deviceName: print(f"Device Disconnected: {deviceName}"))

error = tracker.connectToDeviceOnComPort(COMPORT)
# Check if connection was successful
if error.value() != AisErrorCode.Success:
    print(error.message())
    sys.exit()
    
# Calling sys.exit(app.exec()) will keep the program running until the application is exited
sys.exit(app.exec())

File: SquidstatLibrary/examples/C++/dataOutput/dataOutput.cpp
/**
* \example dataOutput.cpp
* This example shows how to capture data from the device, and write it to a CSV file using Qt's QFile and QTextStream classes.
*/

#include "AisExperiment.h"
#include "AisDeviceTracker.h"
#include "AisInstrumentHandler.h"
#include "experiments/builder_elements/AisConstantCurrentElement.h"
#include "experiments/builder_elements/AisConstantPotElement.h"

#include <QCoreApplication>
#include <QTimer>
#include <QDebug>
#include <QFile>
#include <QStandardPaths>

// Define relevant device information, for easy access
#define COMPORT "COM1"
#define CHANNEL 0

int main()
{
    char** test = nullptr;
    int args;
    QCoreApplication a(args, test);

    auto tracker = AisDeviceTracker::Instance(); 

    // Build the experiment
    AisConstantPotElement cvElement(
        1,      // voltage: 1v
        1,      // sampling interval: 1s
        30      // duration: 30s
    );
    AisConstantCurrentElement ccElement(
        0.001,  // current: 1mA
        1,      // sampling interval: 1s
        60      // duration: 60s
    );
    auto customExperiment = std::make_shared<AisExperiment>();
    customExperiment->appendElement(ccElement, 1);
    customExperiment->appendElement(cvElement, 1);

    // Static QString variable to store the current elements file path
    static QString filePath;

    // Create a lambda funnction to connect signals to the handler
    auto connectSignals = [=](const AisInstrumentHandler& handler) {
        QObject::connect(&handler, &AisInstrumentHandler::experimentNewElementStarting, [=](uint8_t channel, const AisExperimentNode& node) {
            // Create a unique file name for each element, with the format "stepNumber_stepName_expStartTime.csv"
            static int fileNum = 1;
            auto name = "/" + QString::number(fileNum) + "_" + QString::number(node.stepNumber) + "_" + node.stepName + ".csv";
            filePath = QString(QStandardPaths::writableLocation(QStandardPaths::DesktopLocation)) + name;

            QFile file(filePath);
            if (!file.open(QIODevice::WriteOnly | QIODevice::Text))
                return;

            // Writing headers to file
            QTextStream out(&file);
            out << "Time Stamp,"
                << "Counter Electrode Voltage,"
                << "Working Electrode Voltage,"
                << "Current"
                << "\n";
            file.close();

            qDebug() << "New element beginning: " << node.stepName << "step: " << node.stepNumber;
        });

        QObject::connect(&handler, &AisInstrumentHandler::activeDCDataReady, [=](uint8_t channel, const AisDCData& data) {
            qDebug() << "current :" << data.current << "   voltage: " << data.workingElectrodeVoltage << "   counter electrode : " << data.counterElectrodeVoltage << "  timestamp : " << data.timestamp;

            // Save the DC data to the file created at element beginning
            QFile file(filePath);
            if (!file.open(QIODevice::Append | QIODevice::WriteOnly | QIODevice::Text))
                return;
            QTextStream out(&file);
            out << data.timestamp << ","
                << data.counterElectrodeVoltage << ","
                << data.workingElectrodeVoltage << ","
                << data.current
                << "\n";
            file.close();
        });

        // The experiment created only uses DC elements, so this will not be executed.
        // However, ac data can be handled and written to a csv like seen above with dc data
        QObject::connect(&handler, &AisInstrumentHandler::activeACDataReady, [=](uint8_t channel, const AisACData& data) {
            qDebug() << data.frequency << "         " << data.absoluteImpedance << "        " << data.phaseAngle;
        });

        QObject::connect(&handler, &AisInstrumentHandler::experimentStopped, [=](uint8_t channel, const QString& reason) {
            qDebug() << "Experiment Completed Signal " << channel << "Reason : " << reason;
        });
    };

    // When device is connected, setup connections, and upload/start the experiment
    QObject::connect(tracker, &AisDeviceTracker::newDeviceConnected, [=](const QString& deviceName) {
        auto& handler = tracker->getInstrumentHandler(deviceName);

        connectSignals(handler);

        AisErrorCode error = handler.uploadExperimentToChannel(0, customExperiment);
        if (error) {
            qDebug() << error.message();
            return 0;
        }

        error = handler.startUploadedExperiment(CHANNEL);
        if (error) {
            qDebug() << error.message();
            return 0;
        }
    });

    AisErrorCode error = tracker->connectToDeviceOnComPort(COMPORT);
    if (error != error.Success) {
        qDebug() << error.message();
        return 0;
    }

    return a.exec();
}